### PR TITLE
fix(consolidation): run consumer summarize/embed off the write lock (read→compute→write)

### LINKED ISSUE #409: [super-qa] medium: SQLite write transaction held open across consumer LLM/network calls (DoS / lock starvation)
> ⚠️ Generated by `/super-qa` 2026-06-01. Severity & details from agent analysis (security lens ran via octo tier-2 fallback, xhigh not effort:max); **not independently source-verified**. Review before action.

**SQLite write transaction held open across consumer LLM/network calls (DoS / lock starvation)**

- **ID:** `consolidation/dos-lock-held-across-llm-io`
- **Severity:** medium
- **Area:** consolidation · **Type:** security
- **Location:** `src/consolidation/mod.rs:50; src/consolidation/cluster.rs:70-71; src/consolidation/global.rs:59-60`
- **Auto-fixable:** no

## Description
consolidate() opens a single write transaction at mod.rs:50 (conn.unchecked_transaction()) and keeps it open for the entire 3-pass pipeline. Inside that transaction, cluster_fusion calls generator.summarize(&cluster_facts)? and generator.embed(&summary_text)? (cluster.rs:70-71), and global_integration does the same (global.rs:59-60). Per the crate's own design contract, SummaryGenerator delegates ALL LLM/network work to the consumer, so these are unbounded-latency network round-trips. The SQLite write lock is held for the full duration. The engine wraps its write connection in a Mutex (verified: src/engine/mod.rs:508 `fn write_conn(&self) -> Result<MutexGuard<'_, Connection>>`), so a slow, timing-out, or hanging consumer endpoint serializes and stalls the entire engine's write path — every concurrent writer blocks for the round-trip duration. WAL mode keeps readers alive but does not help writers. A consumer with N clusters makes N sequential blocking network calls under the lock. This is the only finding in scope with a real availability attack surface.

## Suggested approach
Do not hold the write transaction across consumer trait calls. Restructure each pass to: (1) read fact/summary inputs under a short read scope, (2) drop the lock and invoke generator.summarize/embed with no transaction held, (3) re-acquire and open a short write transaction only to insert the produced summaries. Example for cluster_fusion: collect all (cluster_facts, source_ids, scope_id) first, release the tx, run all summarize/embed calls, then open a fresh tx for the batch of summary_store.insert(...) calls. Optionally document a consumer-side timeout requirement on SummaryGenerator. Note this trades the current single-transaction atomicity for per-batch atomicity — acceptable since consolidation is idempotent (delete_by_level + rebuild).

## References
CWE-400, CWE-833, CWE-667

### LINKED ISSUE #659: refactor(consolidation): short-circuit dedup/cluster safety cap with COUNT(*) before loading active facts
## Finding

`local_dedup` (and `cluster_fusion`) call `fact_store.list_active(None)` — loading **every** active fact's content, metadata JSON, and high-dimensional embedding into memory — and only *then* check the `> max_facts` safety cap. On the over-cap path this deserializes the entire corpus (~147 MB for 50 000 × 768-dim) purely to discover it should skip.

A lightweight `SELECT COUNT(*)` (active facts) before the full load would let the cap short-circuit **before** paying the deserialization cost.

## Provenance / scope

Surfaced by the gemini-code-assist bot on PR #658 (Wave 2 of the #253 sweep, `dedup.rs:82`). Filed separately rather than folded into #658 per the collateral-issue policy — it is orthogonal to #272's sentinel→`DedupOutcome` refactor.

## Relation to other work

- Overlaps **#389** (active facts loaded twice) and the **Wave 5 (#409/#389) pipeline restructure**, which reworks how/when active facts are loaded across passes. The COUNT(*)-pre-check should be folded into that restructure rather than bolted on independently — implement there if not done sooner.

## Severity

Low: the over-cap path triggers only at ≥ 50 000 active facts (rare for this embedded engine). gemini rated it "high"; downgraded to low given the rare trigger, but the waste-when-triggered is real.

### PR DESCRIPTION
**Pull request by dutiona**
## What & why

`MemoryEngine::consolidate` held the engine's single `write_conn` mutex across the **entire** 3-pass consolidation pipeline — including the consumer `SummaryGenerator::summarize` / `EmbeddingProvider::embed` calls, which are unbounded network IO. While held, every other engine writer — and, for an in-memory pool, every *reader* (reads serialize through the same mutex) — is starved for the full IO duration: a denial-of-service / lock-starvation surface (**#409**, `type:security`, `severity:medium`).

## Approach — read → compute → write (locked decision D3: preserve atomicity)

The pipeline is split so the lock is never held across consumer IO, while keeping all-or-nothing atomicity:

1. **`load_snapshot`** (brief lock): `COUNT` active (#659 short-circuit) + `last_consolidated_at` watermark + `list_active` once (#389).
2. **`compute_plan`** (NO lock): the dedup decision + cluster/global summaries — *all the consumer IO* — computed against the snapshot into a pure `ConsolidationPlan`.
3. **`apply_plan`** (brief lock, ONE transaction): every write applied atomically.

A consumer failure now aborts in the lock-free compute phase, **before any write** — strictly stronger than the prior tx-rollback (`failing_pass_rolls_back_entire_transaction` still holds). The single-connection `consolidate`/per-pass wrappers are retained as `#[cfg(test)]` helpers; the production path is compute + apply.

## Folds in #659

`FactStore::count_active` lets `load_snapshot` short-circuit the expensive `list_active` load (which deserializes every embedding BLOB) when the corpus is over **both** safety caps — both passes would skip anyway.

## Concurrency robustness (read→write gap)

Releasing the lock between snapshot and apply lets a concurrent writer (`prune`, conflict resolution, dream cycle) expire a planned-for fact in the gap. `apply_dedup` handles both cases and returns the ids it **actually** expired (so `notify_expire`/graph-rebuild/stats key off reality, not the stale plan):
- **loser concurrently expired** → tolerated (`NotFound` no-op), not counted;
- **survivor concurrently expired** → the merge is void, so the loser is **kept** rather than expired-too (no orphaned duplicate group). Survivor liveness is snapshotted *before* any expiry so merge **chains** stay correct.

## Intentional behavior change (surfaced, not buried)

When the **cluster** pass is skipped over its cap, global integration no longer re-summarizes the *stale* store clusters — summary writes are gated on `cluster_ran`, so cluster + global are only rewritten when clustering actually runs. Cleaner invariant; this edge case was untested (the cluster cap is 50k and not test-injectable on the old path).

## Test plan

- [x] `cargo build --workspace --all-features`
- [x] `cargo test --workspace` (1538 passed, 9 ignored)
- [x] `cargo clippy --workspace --all-targets --all-features -- -D warnings` (clean)
- [x] `cargo fmt --check`
- [x] `cargo deny check` (advisories/bans/licenses/sources ok)
- [x] New tests: deterministic lock-free probe (a generator that writes through the engine from inside `summarize` — pre-fix it tripped the pool's non-reentrant guard); `count_active`; over-both-caps no-op; concurrent-loser tolerance; concurrent-survivor keeps-loser.

## Review

3 adversarial-lens subagent review (behavior-preservation / concurrency-atomicity / API-blast-radius). Behavior-preservation: LGTM. The substantive concurrency finding (survivor-liveness) is fixed here; the API findings (broken doc-link, stale narrative doc, over-visible helper) are addressed.

Closes #409.
Closes #659.

### CODE REVIEW

File: src/consolidation/mod.rs [hunk 1]
```diff
@@ -135,93 +155,267 @@ fn consolidate_with_caps(
...
+    // Dedup writes: importance inheritances + expirations (concurrency-tolerant, #409).
+    // Returns the ids actually expired by this call.
+    let expired = dedup::apply_dedup(&tx, embed_dim, &plan.dedup, plan.now)?;
+
+    // Summary writes — only when clustering actually ran (#345): over the cap we must not
+    // delete existing summaries without replacements. Cluster + global move together so
+    // global never re-summarizes stale clusters.
+    if plan.cluster_ran {
+        cluster::apply_clusters(&tx, embed_dim, &plan.cluster_summaries)?;
+        global::apply_global(&tx, embed_dim, plan.global_summary.as_ref())?;
```
[the remark this answers is not in the document]
Comment (dutiona) [reply src/consolidation/mod.rs:334]: Fixed in 440f699. `apply_dedup` now returns `DedupApplied { expired, survivor_lost }`; `survivor_lost` is set only when a planned loser is kept because its survivor was concurrently expired (the loser-`NotFound` case keeps it false — those summaries stay valid). `apply_plan` gates the cluster + global summary writes, the embedding-identity stamp, and the `clusters_created`/`global_summaries` stats on `!survivor_lost`: on a loss it preserves the existing summaries and lets the next consolidation rebuild from the corrected active set. Added `apply_plan_skips_stale_summary_writes_when_survivor_lost`. Good catch.

File: src/consolidation/mod.rs [hunk 2]
```diff
@@ -64,66 +76,74 @@ pub fn summarize_and_embed(
...
+    /// New cluster summaries to write (already summarized + embedded).
+    cluster_summaries: Vec<NewSummary>,
+    /// New global summary to write, if any.
+    global_summary: Option<NewSummary>,
+    /// The embedder's fingerprint to stamp, set **iff** a summary vector was produced
+    /// (#643). Captured during compute so `apply_plan` needs no embedder.
+    embedding_fingerprint: Option<EmbeddingFingerprint>,
+    /// Run-level timestamp (expiry stamp + watermark).
+    now: DateTime<Utc>,
+}
```
[the remark this answers is not in the document]
Comment (dutiona) [reply src/consolidation/mod.rs:113]: Declining — false positive. The only code that reads `ConsolidationPlan` (and its fields) is in `crate::consolidation::tests` (the `#[cfg(test)] mod tests` inside `consolidation/mod.rs`), a **descendant** of `consolidation` that can access `pub(super)`/private items. Nothing in `crate::engine::tests` — or anywhere outside `src/consolidation/` — references `ConsolidationPlan`; verified by grep and confirmed by the green build + 1539 passing tests (a real E0616 would not compile). Widening to `pub(crate)` would also reintroduce a `clippy::redundant_pub_crate` error (the `consolidation` module is itself `pub(crate)`), and CI runs `-D warnings`.

File: src/consolidation/dedup.rs [hunk 3]
```diff
@@ -49,53 +42,90 @@ pub enum DedupOutcome {
...
+/// merge decision is void and the `loser` must be kept (not expired), so the duplicate
+/// group is never left without an active representative. `survivor` is the **immediate**
+/// survivor of the pairwise decision; chain correctness comes from snapshotting every
+/// survivor's liveness in `apply_dedup` *before* any expiry (see there).
+pub(super) struct Expiry {
+    /// The lower-importance fact to expire.
+    pub(super) loser: i64,
+    /// The fact `loser` was folded into (the higher-importance member of the pair).
+    pub(super) survivor: i64,
+}
```
[the remark this answers is not in the document]
Comment (dutiona) [reply src/consolidation/dedup.rs:58]: Declining — false positive. The only code that reads `Expiry` (and its fields) is in `crate::consolidation::tests` (the `#[cfg(test)] mod tests` inside `consolidation/mod.rs`), a **descendant** of `consolidation` that can access `pub(super)`/private items. Nothing in `crate::engine::tests` — or anywhere outside `src/consolidation/` — references `Expiry`; verified by grep and confirmed by the green build + 1539 passing tests (a real E0616 would not compile). Widening to `pub(crate)` would also reintroduce a `clippy::redundant_pub_crate` error (the `consolidation` module is itself `pub(crate)`), and CI runs `-D warnings`.

File: src/consolidation/dedup.rs [hunk 4]
```diff
@@ -49,53 +42,90 @@ pub enum DedupOutcome {
...
+    /// Planned expiries (loser + its survivor), the lower-importance member of each
+    /// duplicate pair paired with the fact it merges into.
+    pub(super) expirations: Vec<Expiry>,
+    /// Base-`importance` inheritances `(survivor_id, importance)` — empty under the
+    /// current expiry rule (kept for write-set fidelity; see the struct docs).
+    pub(super) importance_updates: Vec<(i64, f64)>,
+    /// `importance_score` inheritances `(survivor_id, score)` — the #264 running-max,
+    /// one final entry per survivor that absorbed a strictly-higher score.
+    pub(super) importance_score_updates: Vec<(i64, f64)>,
+}
```
[the remark this answers is not in the document]
Comment (dutiona) [reply src/consolidation/dedup.rs:84]: Declining — false positive. The only code that reads `DedupComputed` (and its fields) is in `crate::consolidation::tests` (the `#[cfg(test)] mod tests` inside `consolidation/mod.rs`), a **descendant** of `consolidation` that can access `pub(super)`/private items. Nothing in `crate::engine::tests` — or anywhere outside `src/consolidation/` — references `DedupComputed`; verified by grep and confirmed by the green build + 1539 passing tests (a real E0616 would not compile). Widening to `pub(crate)` would also reintroduce a `clippy::redundant_pub_crate` error (the `consolidation` module is itself `pub(crate)`), and CI runs `-D warnings`.

**Comment by dutiona** [comment]
📋 Plan archived as #721 (`type:plan`). Collateral (pre-existing, out of scope) filed as #720.
